### PR TITLE
feat: add healthcheck

### PR DESCRIPTION
**Pull request by olizilla**
adds an http server to the backup process that provides a healthcheck status. We're seeing the backup process just lock up in prod with no logging but also no process failure or other error to help debug. The plan is to wire this in to ECS so that it restarts the container if it fails healthcheck.

if it's been more than 2 mins since last log, assume we are stalled and return 500.

update dockerfile to use curl to test the healthcheck url

License: MIT

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,8 +1,10 @@
...
+FROM node:18.15.0
 
 WORKDIR /usr/src/app
 COPY package*.json *.js ./
 RUN npm ci --omit=dev
 
 COPY *.js *.json *.sh ./
 CMD [ "npm", "start", "--silent"]
+
+HEALTHCHECK --interval=60s --timeout=1s --start-period=5s --retries=3 CMD [ "curl" "https://127.0.0.1:9999" "--fail" ]
```
Comment (olizilla) [inline Dockerfile:10]: this is a serving suggestion. I will be configuring this in ecs / terraform explicitly.

> The Amazon ECS container agent only monitors and reports on the health checks specified in the task definition. Amazon ECS does not monitor Docker health checks that are embedded in a container image and not specified in the container definition. Health check parameters that are specified in a container definition override any Docker health checks that exist in the container image.
– https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_HealthCheck.html

File: Dockerfile [hunk 2]
```diff
@@ -1,8 +1,10 @@
-FROM node:18.14.2-slim
+FROM node:18.15.0
```
Comment (gobengo) [inline Dockerfile:1]: I assume dropping `-slim` was required to use `curl`?

File: index.js [hunk 3]
```diff
@@ -31,9 +32,18 @@ const REPORT_INTERVAL = 1000 * 60 // log download progress every minute
...
   const sourceDataFile = dataURL.substring(dataURL.lastIndexOf('/') + 1)
-  const log = debug(`backup:${sourceDataFile}`)
+  const logger = debug(`backup:${sourceDataFile}`)
+  const log = (msg) => {
+    logger(msg)
+    health.updateLastLogged()
+  }
+  const gracePeriod = REPORT_INTERVAL * 2
+  const health = createHealthCheckServer({ sourceDataFile, gracePeriod })
+  health.srv.listen(healthcheckPort, '127.0.0.1', () => {
```
Comment (gobengo) [inline index.js:44]: is it important to listen on ipv4 vs `'localhost'`?

**Review by gobengo — APPROVED** [review]
seems like a great idea for the Dockerfile to have a `HEALTHCHECK`

File: Dockerfile (hunk 2 above)
Comment (olizilla) [reply Dockerfile:1]: honestly i didn't check. I just assumed that if `slim` is to have meaning, it wouldn't include things like `curl`... but i was wrong.

- node:18.15.0 is a descendant of bullseye-slim https://hub.docker.com/layers/library/node/18.15.0-slim/images/sha256-e9ea1b9c6ed95b612a2766b7d9399ba35a8cf8068558ca631e08e19c6d6db92d?context=explore
- and the node bullseye-slim 18 dockerfile explicitly installs `curl`! https://github.com/nodejs/docker-node/blob/3c4fa6daf06a4786d202f2f610351837806a0380/18/bullseye-slim/Dockerfile#L20

Reverting to slim!

File: index.js (hunk 3 above)
Comment (olizilla) [reply index.js:44]: mmm `localhost` on my machine defaults to ipv6 which then causes me issue else where so i'm conditioned to be specific in these things. I know the target machines are not yet configured for ipv6.

**Review by alanshaw — APPROVED** [review]

File: test/health.test.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        t.true(obj.msSinceLastLog > gracePeriodMs, JSON.stringify(obj))
+
+        health.updateLastLogged()
+
+        res = await fetch('http://127.0.0.1:9999')
+        t.is(res.status, 200)
+        obj = await res.json()
+        t.true(obj.msSinceLastLog <= gracePeriodMs)
+
+        resolve()
```
[the remark this answers is not in the document]
Comment (olizilla) [reply test/health.test.js:41]: wise

**Comment by olizilla** [comment]
Added a check to set health ok permanently once completed, otherwise we'd force completed containers to keep restarting.

I don't know why the existing completed tasks aren't stopping once completed tho... we see them log `backup complete 🎉` but they persist as a live task in the backup service. This is kinda handy so we can see what's going on, but i think ideally we'd be shutting them down as they complete.

This PR adds an http server, so they are definitely not gonna shutdown automatically now, but that's easy to fix if we want it. The bind here is that until i figure our why they are not cleanly shutting down i **have** to leave the healthcheck http server running, otherwise they will be marked as unhealthy once completed and restarted due to the healthcheck failing.

So, as the tasks currently don't shutdown, i am making it so they are healthy once completed and getting this rolled out.

**Pull request merged**